### PR TITLE
Sot v3 - Making the code independent from HRP-2

### PR DESCRIPTION
**Pull request by olivier-stasse**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -37,18 +37,14 @@ ENDIF(UNIX)
 
 #This project will create many plugins as shared libraries, listed here
 SET(plugins
-  joint-torque-controller
-  force-torque-estimator
-  joint-trajectory-generator
-  nd-trajectory-generator
-  se3-trajectory-generator
-  inverse-dynamics-controller
+#  joint-torque-controller
```
Comment (andreadelprete) [inline src/CMakeLists.txt]: Why did you remove the joint-torque-controller?

Comment (andreadelprete) [reply src/CMakeLists.txt]: @olivier-stasse

Comment (olivier-stasse) [reply src/CMakeLists.txt]: It is a temporary state. Currently the pull request compiles only
inverse-dynamics-balance-controller
position-controller
control-manager
free-flyer-locator
Because they are the only ones which have been translated (yes, I am late...)

File: src/free-flyer-locator.cpp [hunk 2]
```diff
@@ -155,12 +178,12 @@ namespace dynamicgraph
           m_Mff = se3::SE3(se3::exp3(w), 0.5 * (iMo1.translation()+iMo2.translation() ));
 
           // due to distance from ankle to ground
-          Eigen::Map<const Eigen::Vector3d> righ_foot_sole_xyz(&RIGHT_FOOT_SOLE_XYZ[0]);
+          Eigen::Map<const Eigen::Vector3d> righ_foot_sole_xyz(&m_Right_Foot_Sole_XYZ[0]);
```
Comment (andreadelprete) [inline src/free-flyer-locator.cpp]: I think m_Right_Foot_Sole_XYZ is not initialized

File: src/free-flyer-locator.cpp [hunk 3]
```diff
@@ -155,12 +178,12 @@ namespace dynamicgraph
...
 
-          m_q_sot.tail<N_JOINTS>() = q.tail<N_JOINTS>();
+          m_q_sot.tail(m_nbJoints) = q.tail(m_nbJoints);
           base_se3_to_sot(m_Mff.translation()-righ_foot_sole_xyz,
                           m_Mff.rotation(),
-                          m_q_sot.head<6>());
```
Comment (andreadelprete) [inline src/free-flyer-locator.cpp]: I do not understand why you changed head<6> into head(6)

File: src/inverse-dynamics-balance-controller.cpp [hunk 4]
```diff
@@ -249,6 +249,23 @@ namespace dynamicgraph
                    makeCommandVoid1(*this, &InverseDynamicsBalanceController::removeLeftFootContact,
                                     docCommandVoid1("Remove the contact at the left foot.",
                                                     "Transition time in seconds (double)")));
+
+	addCommand("setLeftFootFrameName",
```
Comment (andreadelprete) [inline src/inverse-dynamics-balance-controller.cpp]: I would take the foot frame names as inputs in the init command

File: src/position-controller.cpp [hunk 5]
```diff
@@ -76,15 +75,17 @@ namespace dynamicgraph
 
         /* Commands. */
         addCommand("init",
-                   makeCommandVoid1(*this, &PositionController::init,
-                                    docCommandVoid1("Initialize the entity.",
-                                                    "Time period in seconds (double)")));
+                   makeCommandVoid2(*this, &PositionController::init,
+                                    docCommandVoid2("Initialize the entity.",
+                                                    "Time period in seconds (double)",
+						    "Size of the state vector (double)")));
```
Comment (andreadelprete) [inline src/position-controller.cpp:82]: I think "size of the state vector" is an ambiguous description. Maybe "number of joints (not including floating base)" is better.

File: unitTesting/test_control_manager.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+max_tau = 100.0*ones(30)
+percentageDriverDeadZoneCompensation = 20.0*ones(30)
+signWindowsFilterSize = ones(30)
+tau = 100.0 * ones(30)
+tau_predicted = 110.0 * ones(30)
+pwmDes = 100.0 *ones(30)
+currentDes = 100.0 *ones(30)
+
+# Setting the robot configuration
+cm.base6d_encoders.value=q
```
Comment (andreadelprete) [inline unitTesting/test_control_manager.py]: Does this mean that in sot v3 you can directly set the value of a signal to a numpy array? Do tuples still work?

File: unitTesting/test_free_flyer_locator.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+from dynamic_graph.sot.core.matrix_util import matrixToTuple, vectorToTuple,rotate, matrixToRPY
+from dynamic_graph.sot.torque_control.control_manager import *
```
Comment (andreadelprete) [inline unitTesting/test_free_flyer_locator.py]: [minor] Probably this line is useless

**Review by andreadelprete — CHANGES_REQUESTED** [review]
All the comments are in the code. The main issue is the mapping of the joint order from urdf to sot that is not properly handled in the control manager, together with the fact that we lose the "short joint names" that I find very practical. All the comments are in the code. Anyway I think we're on the right path and it shouldn't take much more to have something working! 👍

File: include/sot/torque_control/free-flyer-locator.hh [hunk 8]
```diff
@@ -105,21 +110,28 @@ namespace dynamicgraph {
...
+        long unsigned int      m_left_foot_id;
         Eigen::VectorXd   m_q_pin;            /// robot configuration according to pinocchio convention
         Eigen::VectorXd   m_q_sot;            /// robot configuration according to SoT convention
         Eigen::VectorXd   m_v_pin;            /// robot velocities according to pinocchio convention
         Eigen::VectorXd   m_v_sot;            /// robot velocities according to SoT convention
 
+	std::string m_Left_Foot_Frame_Name;
+	std::string m_Right_Foot_Frame_Name;
+	unsigned int m_nbJoints;
+	double m_Right_Foot_Sole_XYZ[3];
```
[the remark this answers is not in the document]
Comment (andreadelprete) [reply include/sot/torque_control/free-flyer-locator.hh]: I just realized that this is how the free-flyer-locator has been implemented, so ignore this comment!

File: src/control-manager.cpp [hunk 9]
```diff
@@ -71,23 +72,19 @@ namespace dynamicgraph
             ,CONSTRUCT_SIGNAL_OUT(pwmDesSafe,dynamicgraph::Vector, INPUT_SIGNALS << m_pwmDesSOUT)
             ,m_initSucceeded(false)
             ,m_emergency_stop_triggered(false)
-            ,m_maxCurrent(DEFAULT_MAX_CURRENT)
+            ,m_maxCurrent(5)
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: The init command policy is now implemented in commit 7f142f2

File: src/control-manager.cpp [hunk 10]
```diff
@@ -71,23 +72,24 @@ namespace dynamicgraph
...
             ,m_is_first_iter(true)
       {
-        m_jointCtrlModes_current.resize(N_JOINTS);
-        m_jointCtrlModes_previous.resize(N_JOINTS);
-        m_jointCtrlModesCountDown.resize(N_JOINTS,0);
-        m_signIsPos.resize(N_JOINTS, false);
-        m_changeSignCpt.resize(N_JOINTS, 0);
-        m_winSizeAdapt.resize(N_JOINTS, 0);
+        m_jointCtrlModes_current.resize(m_nJoints);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fixed in 8ccc3cc73437bd5d94f3ee830f65f6190f4292e0

File: src/control-manager.cpp [hunk 11]
```diff
@@ -256,7 +274,7 @@ namespace dynamicgraph
             {
               m_emergency_stop_triggered = true;
               SEND_MSG("Estimated torque "+toString(tau(i))+" > max torque "+toString(tau_max(i))+
-                       " for joint "+JointUtil::get_name_from_id(i), MSG_TYPE_ERROR);
+                       " for joint "+ m_robot->model().getJointName(i), MSG_TYPE_ERROR);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit bc5f539678009aa87d3eaa2a26e524133f5d03dd. But IMHO adding a remapping between name and id in a complex robotics application, in the long run, may just generate another set of confusion for newcomers...

File: src/control-manager.cpp [hunk 12]
```diff
@@ -276,14 +294,12 @@ namespace dynamicgraph
...
               m_maxCurrent = m_max_currentSIN(iter)(i);
-            else
-              m_maxCurrent = DEFAULT_MAX_CURRENT;
 
             if( (fabs(pwmDes(i)) > m_maxCurrent) || 
                 (fabs(s(i))      > m_maxCurrent * FROM_CURRENT_TO_12_BIT_CTRL) )
             {
               m_emergency_stop_triggered = true;
-              SEND_MSG("Joint "+JointUtil::get_name_from_id(i)+" desired current is too large: "+
+              SEND_MSG("Joint "+m_robot->model().getJointName(i)+" desired current is too large: "+
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit bc5f539678009aa87d3eaa2a26e524133f5d03dd. But IMHO adding a remapping between name and id in a complex robotics application, in the long run, may just generate another set of confusion for newcomers...

File: src/control-manager.cpp [hunk 13]
```diff
@@ -417,7 +433,7 @@ namespace dynamicgraph
           }
         }
         else
-          SEND_MSG("Cannot change control mode of joint "+JointUtil::get_name_from_id(jid)+
+          SEND_MSG("Cannot change control mode of joint "+m_robot->model().getJointName(jid)+
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit bc5f539678009aa87d3eaa2a26e524133f5d03dd. But IMHO adding a remapping between name and id in a complex robotics application, in the long run, may just generate another set of confusion for newcomers...

File: src/control-manager.cpp [hunk 14]
```diff
@@ -427,8 +443,8 @@ namespace dynamicgraph
         if(jointName=="all")
         {
           stringstream ss;
-          for(int i=0; i<N_JOINTS; i++)
-            ss<<JointUtil::get_name_from_id(i)<<" "<<m_jointCtrlModes_current[i]<<"; ";
+          for(unsigned int i=0; i<m_nJoints; i++)
+            ss<<m_robot->model().getJointName(i)<<" "<<m_jointCtrlModes_current[i]<<"; ";
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit bc5f539678009aa87d3eaa2a26e524133f5d03dd. But IMHO adding a remapping between name and id in a complex robotics application, in the long run, may just generate another set of confusion for newcomers...

File: src/control-manager.cpp [hunk 15]
```diff
@@ -445,21 +461,32 @@ namespace dynamicgraph
         getStatistics().reset_all();
       }
 
+      void ControlManager::setDefaultMaxCurrent( const double & lDefaultMaxCurrent)
+      {
+	m_maxCurrent = lDefaultMaxCurrent;
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit b5dce19c92a02dab296de149491bb0576be287c1

File: src/control-manager.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreadelprete) [reply src/control-manager.cpp]: I'd love to avoid this mapping, but to do so we should have the same joint order in the SoT and in the URDF. Is that possible for you?

File: src/control-manager.cpp [hunk 16]
```diff
@@ -500,15 +527,16 @@ namespace dynamicgraph
 
       bool ControlManager::isJointInRange(unsigned int id, double q)
       {
-        JointLimits jl = JointUtil::get_limits_from_id(id);
-        if(q<jl.lower)
+	double jl= m_robot->model().lowerPositionLimit[id];
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit ce97337c7e0c47f51f1b46db9228837af97b673d

File: src/control-manager.cpp [hunk 17]
```diff
@@ -500,15 +527,16 @@ namespace dynamicgraph
...
+        if(q<jl)
         {
-          SEND_MSG("Desired joint angle "+toString(q)+" is smaller than lower limit: "+toString(jl.lower),MSG_TYPE_ERROR);
+          SEND_MSG("Desired joint angle "+toString(q)+" is smaller than lower limit: "+toString(jl),MSG_TYPE_ERROR);
           return false;
         }
-        if(q>jl.upper)
+	double ju = m_robot->model().upperPositionLimit[id];
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp]: Fix in commit ce97337c7e0c47f51f1b46db9228837af97b673d

File: include/sot/torque_control/free-flyer-locator.hh [hunk 18]
```diff
@@ -79,8 +80,12 @@ namespace dynamicgraph {
         /* --- CONSTRUCTOR ---- */
         FreeFlyerLocator( const std::string & name );
 
-        void init(const std::string& urdfFile);
+        void init(const std::string& urdfFile,
+		  const std::string & leftFootFrameName,
+		  const std::string & rightFootFrameName);
 
+	/// Set the mapping between urdf and sot.
+	void setJoints(const dynamicgraph::Vector &);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply include/sot/torque_control/free-flyer-locator.hh]: It is not obvious to translate the mapping in something easily understandable by the SoT type caster. Could you have a look to test_control_manager.py:33 for an alternative and give some feedback ? Thanks for your help.

File: include/sot/torque_control/free-flyer-locator.hh [hunk 19]
```diff
@@ -79,8 +80,12 @@ namespace dynamicgraph {
...
         void resetIntegral();
```
[the remark this answers is not in the document]
Comment (thomasfla) [reply include/sot/torque_control/free-flyer-locator.hh]: I do confirm!

**Comment by olivier-stasse** [comment]
This pull request has been superseded by https://github.com/stack-of-tasks/sot-torque-control/pull/22

**Review by andreadelprete — COMMENTED** [review]
A few comments here and there

File: include/sot/torque_control/control-manager.hh [hunk 20]
```diff
@@ -126,14 +126,36 @@ namespace dynamicgraph {
...
         void getCtrlMode(const std::string& jointName);
         void setCtrlMode(const std::string& jointName, const std::string& ctrlMode);
         void setCtrlMode(const int jid, const CtrlMode& cm);
+	
         void resetProfiler();
 	void setDefaultMaxCurrent(const double &lDefaultMaxCurrent);
+	
+	/// Commands related to joint name and joint id
 	void setNameToId(const std::string& jointName, const double & jointId);
+	void setJointLimitsFromId(const double &jointId,
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply include/sot/torque_control/control-manager.hh]: Just sticking to the current behavior.
Will do it later.

File: include/sot/torque_control/control-manager.hh [hunk 21]
```diff
@@ -126,14 +126,36 @@ namespace dynamicgraph {
...
+
+	/// Command related to ForceUtil
+	void setForceLimitsFromId(const double &jointId, 
+				  const dynamicgraph::Vector &lq, 
+				  const dynamicgraph::Vector &uq);
+	void setForceNameToForceId(const std::string& forceName, 
+				   const double & forceId);
+	
+	/// Commands related to FootUtil
+	void setRightFootSoleXYZ(const dynamicgraph::Vector &);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply include/sot/torque_control/control-manager.hh:153]: Because the other one was not implemented in the original version

File: unitTesting/test_control_manager.py (hunk 6 above)
Comment (proyan) [reply unitTesting/test_control_manager.py]: Tuples work in v3 as well.

File: src/common.cpp [hunk 22]
```diff
@@ -124,8 +291,141 @@ namespace dynamicgraph
 	  q_urdf[idx]=q_sot[m_urdf_to_sot[idx]];	
 	return true;
       }
+      
+      bool RobotUtil::
+      velocity_urdf_to_sot(Eigen::ConstRefVector v_urdf, Eigen::RefVector v_sot)
+      {
+	assert(q_urdf.size()==m_nbJoints+6);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/common.cpp:298]: Just for your information this pull request is closed.
I rebased the code on the latest commit from proyan and this is fixed in
https://github.com/stack-of-tasks/sot-torque-control/pull/22

File: src/common.cpp [hunk 23]
```diff
@@ -124,8 +291,141 @@ namespace dynamicgraph
...
+        v_sot.head<6>() = v_urdf.head<6>();
+        joints_urdf_to_sot(v_urdf.tail(m_nbJoints), 
+			   v_sot.tail(m_nbJoints));
+	return true;
+      }
+
+      bool RobotUtil::
+      velocity_sot_to_urdf(Eigen::ConstRefVector v_sot, Eigen::RefVector v_urdf)
+      {
+	assert(q_urdf.size()==m_nbJoints+6);
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/common.cpp:316]: same thing fixed in https://github.com/stack-of-tasks/sot-torque-control/pull/22

File: src/control-manager.cpp [hunk 24]
```diff
@@ -125,6 +125,47 @@ namespace dynamicgraph
...
 						    "(double) joint id")));
 
+	addCommand("setForceNameToForceId",
+		   makeCommandVoid2(*this,&ControlManager::setForceNameToForceId,
+				    docCommandVoid2("Set map from a force sensor name to a force sensor Id",
+						    "(string) force sensor name",
+						    "(double) force sensor id")));
+
+
+	addCommand("setJointLimitsFromId",
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply src/control-manager.cpp:135]: we will...
